src/contract/event.rs

```diff
@@ -516,12 +544,103 @@ impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
         self
     }
 
+    /// Returns a web3 provider and filter needed for querying and streaming
+    /// events.
+    fn prepare(self) -> (Web3<T>, FilterBuilder) {
+        let mut filter_builder = self.filter.topic_filter(self.topics.compat());
+        if let Some(from_block) = self.from_block {
+            filter_builder = filter_builder.from_block(from_block);
+        }
+        if let Some(to_block) = self.to_block {
+            filter_builder = filter_builder.to_block(to_block);
+        }
+
+        (self.web3, filter_builder)
+    }
+
     /// Returns a future that resolves into a collection of events matching the
     /// event builder's parameters.
     pub fn query(self) -> QueryAllFuture<T, E> {
         QueryAllFuture::from_builder(self)
     }
 
+    /// Returns a future that resolves into a collection of events matching the
+    /// event builder's parameters. This method is similar to `query` with the
+    /// notable difference that the logs are fetched in pages by querying
+    /// smaller block ranges specified by `block_page_size` instead of using a
+    /// single query.
+    ///
+    /// # Panics
+    ///
+    /// If the block range is invalid for querying past events:
+    /// - the from block is "latest" or "pending"
+    /// - the to block is "earliest"
+    pub async fn query_past_events_paginated(self) -> Result<Vec<Event<E>>, ExecutionError> {
+        let mut start_block = match self.from_block {
+            None | Some(BlockNumber::Earliest) => 0,
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                panic!("invalid from block value 'latest' or 'pending'")
+            }
+        };
+        if let Some(deployment_tx) = self.deployment_transaction {
+            start_block = cmp::max(
+                start_block,
+                block_number_from_transaction_hash(self.web3.clone(), deployment_tx).await?,
+            );
+        }
+
+        let end_block = match self.to_block {
+            None | Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                self.web3.eth().block_number().compat().await?.as_u64()
```

Comment:
It does, the last page is queried with `latest` or `pending` depending on what it was here.

Reply:
In that case, shouldn't it be valid to query from Latest to Pending as they are different? 
Maybe it would be simpler to not support Pending blocks for now? Does the event stream for future events allow to operate on pending blocks?

Reply:
That should be valid right now since when `from_block` is `Latest` we just send a single query to the node.

I see your point though, if we use the block number for latest, then this could be an issue. I'll change it so if the range doesn't make sense to the method it just does a single query with the node. That way we get consistent behaviour with whatever the node does for these weird ranges.